text/0000-dyn-error-generic-member-access.md

```diff
@@ -0,0 +1,707 @@
+    }
+
+    /// Tag a Box of a concrete type with a given `Tag`.
+    ///
+    /// This is like an unsizing coercion, but must be performed explicitly to
+    /// specify the specific tag.
+    #[cfg(feature = "alloc")]
+    pub fn tag_box<I>(value: Box<I::Type>) -> Box<dyn Tagged<'a>>
+    where
+        I: Tag<'a>,
+    {
+        // SAFETY: `TaggedImpl<'a, I>` has the same representation as `I::Type`
+        // due to `#[repr(transparent)]`.
+        unsafe { Box::from_raw(Box::into_raw(value) as *mut TaggedImpl<'a, I>) }
+    }
+
+    /// Returns `true` if the dynamic type is tagged with `I`.
+    #[inline]
+    pub fn is<I>(&self) -> bool
+    where
+        I: Tag<'a>,
+    {
+        self.tag_id() == TypeId::of::<I>()
+    }
+
+    /// Returns some reference to the dynamic value if it is tagged with `I`,
+    /// or `None` if it isn't.
+    #[inline]
+    pub fn downcast_ref<I>(&self) -> Option<&I::Type>
+    where
+        I: Tag<'a>,
+    {
+        if self.is::<I>() {
+            // SAFETY: Just checked whether we're pointing to a
+            // `TaggedImpl<'a, I>`, which was cast to from an `I::Type`.
+            unsafe { Some(&*(self as *const dyn Tagged<'a> as *const I::Type)) }
+        } else {
+            None
+        }
+    }
+
+    /// Returns some reference to the dynamic value if it is tagged with `I`,
+    /// or `None` if it isn't.
+    #[inline]
+    pub fn downcast_mut<I>(&mut self) -> Option<&mut I::Type>
+    where
+        I: Tag<'a>,
+    {
+        if self.is::<I>() {
+            // SAFETY: Just checked whether we're pointing to a
+            // `TaggedImpl<'a, I>`, which was cast to from an `I::Type`.
+            unsafe { Some(&mut *(self as *mut dyn Tagged<'a> as *mut I::Type)) }
+        } else {
+            None
+        }
+    }
+
+    #[inline]
+    #[cfg(feature = "alloc")]
+    pub fn downcast_box<I>(self: Box<Self>) -> Result<Box<I::Type>, Box<Self>>
+    where
+        I: Tag<'a>,
+    {
+        if self.is::<I>() {
+            unsafe {
+                // SAFETY: Just checked whether we're pointing to a
+                // `TaggedImpl<'a, I>`, which was cast to from an `I::Type`.
+                let raw: *mut dyn Tagged<'a> = Box::into_raw(self);
+                Ok(Box::from_raw(raw as *mut I::Type))
+            }
+        } else {
+            Err(self)
+        }
+    }
+}
+```
+
+### Define a generic accessor on the `Error` trait
+
+```rust
+pub trait Error {
+    // ...
+
+    fn provide_context<'a>(&'a self, _request: &mut Request<'a>) {}
+}
+```
+
+### Use this `Request` type to handle passing generic types out of the trait object
+
+```rust
+impl dyn Error {
+    pub fn context_ref<T: ?Sized + 'static>(&self) -> Option<&T> {
+        Request::request_ref(|req| self.provide_context(req))
+    }
+
+    pub fn context<T: 'static>(&self) -> Option<T> {
+        Request::request_value(|req| self.provide_context(req))
+    }
+}
+```
+
+# Drawbacks
```

Comment:
A potential additional drawback is that requesting types dynamically like this can lead to surprising changes in runtime behavior when changing versions of (transitive) dependencies.

Given:

* A crate `foo` with versions 1.0.0 and 2.0.0, each of which provides a type `Foo`.
* A crate `bar` which depends on version 1.0.0 of `foo` and whose errors provide `Foo`.
* A crate `baz` which depends `bar` and on version 2.0.0 of `foo` and tries to request `Foo` from `bar`'s errors.

Then a situation is created in which `baz` will fail to request `Foo` from `bar`'s error because the requested `Foo` and provided `Foo` are actually distinct types because the transitive dependency `foo`'s versions don't match up.

This could even happen in a semver-compatible release of `bar` and `baz` because technically the API is not broken: the type can still be requested, but it will simply begin returning `None` instead of `Some(_)`. This would be even worse if `foo`, `bar,` and `baz` are all library crates depended on by an application crate and the semver-incompatible versions of `foo` are introduced by a semver-compatible change in `bar` or `baz` which would get pulled in by running `cargo update` on the application crate. The application crate has no recourse in this situation aside from waiting for `bar` and `baz` to agree on a semver-compatible version of `foo` again.

(I've been playing with an alternate design based on an associated type rather than generic member access which causes this situation to be a compile time error instead of a runtime behavior change, but that design requires traits with lifetime GATs to be dyn safe to be actually good, and associated type defaults for it to *maybe* be possible to retrofit that design onto the standard library's existing error trait.)

Reply:
I agree this is a footgun I'd prefer to avoid but as far as I can tell this is not a problem being uniquely introduced by generic member access. This is a problem with type erasure in general and the same situation could be encountered by having foo return `Foo` as a `Box<dyn Any>` in any of it's public API's. If `baz` tries to downcast using it's own path to `foo::Foo` it will run into the same type ID mismatch.

I avoid this in my crates currently by re-exporting types from crates that are part of my public API (https://docs.rs/color-eyre/latest/color_eyre/#reexports). Applying this to your example you could avoid this footgun if you downcast/request `Foo` via `bar::foo::Foo` which would correctly resolve to whatever version it is using. This of course gets a little silly when you want to be able to handle all the versions of `Foo`, because then presumably you have to try to request `Foo` once for each path through which you could possibly reach it, which is definitely a subpar user experience, so not necessarily a great solution.

That said, I was never able to figure out how to implement generic member access w/o type erasure and runtime checks. I'm curious to understand more about your potential alternative design, because I would definitely love to see a better solution but I don't see how it is possible given the constraints of the error trait and the goals of generic member access.

Reply:
> I agree this is a footgun I'd prefer to avoid but as far as I can tell this is not a problem being uniquely introduced by generic member access.

True.

> This is a problem with type erasure in general and the same situation could be encountered by having foo return `Foo` as a `Box<dyn Any>` in any of it's public API's.

However, it seems to me that this kind of thing is relatively rare currently, both in the standard library and ecosystem crates. The only standard library exception I can think of is `std::io::Error`, for which I found (part of?) the rationale for in https://github.com/rust-lang/rust/pull/24793#issuecomment-97153138, and I haven't personally encountered any ecosystem crates that make use of this and tell users in documentation that they can/should downcast the inner error to some concrete type.

So, I think making type erasure a central part of the error handling API of the standard library would probably be the main place this kind of problem would come up. I'm not sure the existing occasional cases of downcasting in public APIs can be used to justify a standard library API built entirely on downcasting.

> Applying this to your example you could avoid this footgun if you downcast/request `Foo` via `bar::foo::Foo` which would correctly resolve to whatever version it is using.

While true, I imagine another place this might become a problem is in a library whose purpose is to use generic member access to generate pretty error reports. This hypothetical library might hypothetically pull in the `backtrace` crate for handling backtraces. If any other crate in the dependency tree has errors that provide a backtrace from a different version of the `backtrace` crate, the hypothetical library will not be able to extract them. I don't think this can be solved with re-exports, because that would force the library to have a re-export of every crate that has errors that provide a `backtrace` crate backtrace.

Your next point about requesting the type through each path (including crate version) does provide a solution there, but also as you point out, would not be the most fun to maintain.

> I'm curious to understand more about your potential alternative design, because I would definitely love to see a better solution but I don't see how it is possible given the constraints of the error trait and the goals of generic member access.

I can't promise that this particular idea is any better, but for what it's worth, it goes like this:

```rust
trait Error {
    /// Type that contains details about this error.
    ///
    /// For example, this could be a struct containing a String with help text and a Backtrace.
    type Details<'a> = ()
        where Self: 'a;

    /// Get the details about this error.
    fn details(&self) -> Self::Details<'_>;
}
```

As written, this requires some language features that aren't implemented yet:

1. Associated type defaults. Necessary to make it possible to backport this onto `core::error::Error`, but luckily not required for defining a new error trait in a new crate.
2. `dyn`-safety for traits with lifetime GATs. In my current implementation, I've simply dropped the lifetime and am just returning an owned value, which has some overhead but at least allows experimenting with the rest of the design.

Another part of this design is a `map`-like function to change the type of `Details` when composing errors that have different concrete types.

To approximate generic member access to make it possible to implement a hypothetical library for pretty-printing errors like described above, I imagine a crate could define some traits like `trait GetBacktrace { fn get_backtrace(&self) -> Option<&Backtrace>; }` (or one trait with many methods, I suppose) and then adding these traits as bounds on `Error::Details` on the pretty-printing functions. The author of a binary crate could then implement these traits for their `Error::Details` type and use the library's pretty-printing code.

Reply:
Wouldn't that still have a similar problem? To adapt your example from before:


- A crate foo with versions 1.0.0 and 2.0.0, each of which provides a trait Foo.
- A crate bar which depends on version 1.0.0 of foo and whose errors have a `Details` type that implements Foo 
- A crate baz which depends bar and on version 2.0.0 of foo and has uses a bound that requires `Error::Details` to implement Foo (from version 2.0.0 of foo)

baz still can't get the the `Foo` details from errors from bar.

Reply:
That's correct, but the difference is *how* it can't get the details. With the generic member access design, this would fail at run time: the generic member access functions would return `None`. With the associated type design, this would fail at compile time: a compilation error would be emitted about the trait bound being unsatisfied.